### PR TITLE
翻译：标准库sync

### PR DESCRIPTION
**Pull request by Code-Fight**
[翻译：标准库sync ISSUE](https://github.com/studygolang/website/issues/32)

This PR will be imported into Gerrit with the title and first
comment (this text) used to generate the subject and body of
the Gerrit change.

**Please ensure you adhere to every item in this list.**

More info can be found at https://github.com/golang/go/wiki/CommitMessage

+ The PR title is formatted as follows: `net/http: frob the quux before blarfing`
  + The package name goes before the colon
  + The part after the colon uses the verb tense + phrase that completes the blank in,
    "This change modifies Go to ___________"
  + Lowercase verb after the colon
  + No trailing period
  + Keep the title as short as possible. ideally under 76 characters or shorter
+ No Markdown
+ The first PR comment (this one) is wrapped at 76 characters, unless it's
  really needed (ASCII art, table, or long link)
+ If there is a corresponding issue, add either `Fixes #1234` or `Updates #1234`
  (the latter if this is not a complete fix) to this comment
+ If referring to a repo other than `golang/go` you can use the
  `owner/repo#issue_number` syntax: `Fixes golang/tools#1234`
+ We do not use Signed-off-by lines in Go. Please don't add them.
  Our Gerrit server & GitHub bots enforce CLA compliance instead.
+ Delete these instructions once you have read and applied them

### CODE REVIEW

File: src/sync/atomic/doc.go [hunk 1]
```diff
@@ -2,144 +2,140 @@
...
-// SwapUintptr atomically stores new into *addr and returns the previous *addr value.
+// SwapUintptr 将 new 的原子的存储到 *addr 中，并返回 *addr 之前的值.
 func SwapUintptr(addr *uintptr, new uintptr) (old uintptr)
 
-// SwapPointer atomically stores new into *addr and returns the previous *addr value.
+// SwapPointer 将 new 的原子的存储到 *addr 中，并返回 *addr 之前的值.
 func SwapPointer(addr *unsafe.Pointer, new unsafe.Pointer) (old unsafe.Pointer)
 
-// CompareAndSwapInt32 executes the compare-and-swap operation for an int32 value.
+// CompareAndSwapInt32 适用于Int32的对比并交换；先对比new是否与old相同，如果相同就交换并返回true，如果不同直接返回false.
```
Comment (darcyaf) [inline src/sync/atomic/doc.go]: 这边很多delta空格漏了.
适用于这边的最好都加上空格吧.

File: src/sync/atomic/value.go [hunk 2]
```diff
@@ -8,23 +8,23 @@ import (
 	"unsafe"
 )
 
-// A Value provides an atomic load and store of a consistently typed value.
-// The zero value for a Value returns nil from Load.
-// Once Store has been called, a Value must not be copied.
+// Value 可以原子的存储或加载类型的value.
+// Value 的零值是nil.
+// 一旦使用store保存后, 这个Value进制不允许再次复制.
```
Comment (darcyaf) [inline src/sync/atomic/value.go]: 同上，格式统一一下吧。

File: src/sync/cond.go [hunk 3]
```diff
@@ -9,38 +9,34 @@ import (
...
-	// L is held while observing or changing the condition
+	// 使用Cond的时候必须持有这个锁
 	L Locker
 
 	notify  notifyList
 	checker copyChecker
 }
 
-// NewCond returns a new Cond with Locker l.
+// NewCond 使用传入的l锁新的Cond
```
Comment (darcyaf) [inline src/sync/cond.go]:
```suggestion
// NewCond 使用传入的l锁生成新的 Cond
```

File: src/sync/map.go [hunk 4]
```diff
@@ -9,67 +9,59 @@ import (
...
+// 以提高类型的安全性以及更易于维护.
 //
-// The Map type is optimized for two common use cases: (1) when the entry for a given
-// key is only ever written once but read many times, as in caches that only grow,
-// or (2) when multiple goroutines read, write, and overwrite entries for disjoint
-// sets of keys. In these two cases, use of a Map may significantly reduce lock
-// contention compared to a Go map paired with a separate Mutex or RWMutex.
+// Sync.map 针对两种场景进行了优化
+// (1) 当给定的集合是一个读多写少的场景时,比如缓存
+// (2) 当多个goroutine去读写、甚至覆盖完全不想交的条目时
```
Comment (darcyaf) [inline src/sync/map.go]:
```suggestion
// (2) 当多个goroutine去读写、甚至覆盖完全不相交的条目时
```

File: src/sync/map.go [hunk 5]
```diff
@@ -9,67 +9,59 @@ import (
...
+// 在这两种场景下，Sync.map 比使用原生的map配合锁的方式可以明显的减少锁的竞争
 //
-// The zero Map is empty and ready for use. A Map must not be copied after first use.
+// Map不需要初始化，零值时即可立即使用，但是第一次使用后，不允许复制
 type Map struct {
 	mu Mutex
 
-	// read contains the portion of the map's contents that are safe for
-	// concurrent access (with or without mu held).
+	// 读取read中的
```
Comment (darcyaf) [inline src/sync/map.go]: 这里翻译不完整。

File: src/sync/map.go [hunk 6]
```diff
@@ -9,67 +9,59 @@ import (
...
-	// map and unexpunged with mu held.
-	read atomic.Value // readOnly
+	// read 集合中存储的数据并发更新时不需要持有锁
+	// 如果更新已经删除的条目需要复制到 dirty map中且需要持有锁.
+	read atomic.Value // 只读
 
-	// dirty contains the portion of the map's contents that require mu to be
-	// held. To ensure that the dirty map can be promoted to the read map quickly,
-	// it also includes all of the non-expunged entries in the read map.
+	// dirty 集合的数据条目操作时不需持有锁
```
Comment (darcyaf) [inline src/sync/map.go]:
```suggestion
	// dirty 包含了 需持有锁的数据```

这里逻辑错了。

File: src/sync/map.go [hunk 7]
```diff
@@ -178,24 +168,23 @@ func (e *entry) tryStore(i *interface{}) bool {
 	}
 }
 
-// unexpungeLocked ensures that the entry is not marked as expunged.
+// unexpungeLocked 确保未被标记为删除.
```
Comment (darcyaf) [inline src/sync/map.go]:
```suggestion
// unexpungeLocked 确保条目未被标记为删除.
```

File: src/sync/pool.go [hunk 8]
```diff
@@ -11,55 +11,48 @@ import (
...
+// 如果Pool中只存储了引用，这种情况可能也会被回收掉
 //
-// A Pool is safe for use by multiple goroutines simultaneously.
+// Pool 是并发安全的.
 //
-// Pool's purpose is to cache allocated but unused items for later reuse,
-// relieving pressure on the garbage collector. That is, it makes it easy to
-// build efficient, thread-safe free lists. However, it is not suitable for all
-// free lists.
+// Pool的目的是为了缓存已经创建但为使用的对象，方便后续再次使用, 减轻GC的压力.
```
Comment (darcyaf) [inline src/sync/pool.go]:
```suggestion
// Pool的目的是为了缓存已经创建但未使用的对象，方便后续再次使用, 减轻GC的压力. 
```

File: src/sync/runtime.go [hunk 9]
```diff
@@ -6,52 +6,49 @@ package sync
 
 import "unsafe"
 
-// defined in package runtime
+// 具体的细节请参考 runtime 包
 
-// Semacquire waits until *s > 0 and then atomically decrements it.
-// It is intended as a simple sleep primitive for use by the synchronization
-// library and should not be used directly.
+// Semacquire 当 *s > 0 时，开始去原子的递减.
```
Comment (darcyaf) [inline src/sync/runtime.go]:
```suggestion
// Semacquire 会等待直到 *s > 0 时，开始去原子的递减.
```

File: src/sync/runtime.go [hunk 10]
```diff
@@ -6,52 +6,49 @@ package sync
...
 func runtime_SemacquireMutex(s *uint32, lifo bool, skipframes int)
 
-// Semrelease atomically increments *s and notifies a waiting goroutine
-// if one is blocked in Semacquire.
-// It is intended as a simple wakeup primitive for use by the synchronization
-// library and should not be used directly.
-// If handoff is true, pass count directly to the first waiter.
-// skipframes is the number of frames to omit during tracing, counting from
-// runtime_Semrelease's caller.
+// Semrelease 原子的增加 *s 并且去唤醒因为Semacquire而阻塞等待的 goroutine
```
Comment (darcyaf) [inline src/sync/runtime.go]:
```suggestion
// Semrelease 原子的增加 *s 并且去唤醒因为 Semacquire 而阻塞等待的 goroutine
```

File: src/sync/rwmutex.go [hunk 11]
```diff
@@ -10,36 +10,31 @@ import (
...
-// A RWMutex must not be copied after first use.
+// RWMutex 使用后禁止复制.
 //
-// If a goroutine holds a RWMutex for reading and another goroutine might
-// call Lock, no goroutine should expect to be able to acquire a read lock
-// until the initial read lock is released. In particular, this prohibits
-// recursive read locking. This is to ensure that the lock eventually becomes
-// available; a blocked Lock call excludes new readers from acquiring the
-// lock.
+// 如果goroutine拥有 RWMutex 进行读取，并且另外一个goroutine调用Lock
```
Comment (darcyaf) [inline src/sync/rwmutex.go]:
```suggestion
// 如果 goroutine 拥有 RWMutex 进行读取，并且另外一个 goroutine 调用Lock
```

File: src/sync/rwmutex.go [hunk 12]
```diff
@@ -10,36 +10,31 @@ import (
...
+// 在释放初始化读取锁之前，任何goroutine都不会获取到读取锁
```
Comment (darcyaf) [inline src/sync/rwmutex.go]:
```suggestion
// 在释放初始化读取锁之前，任何 goroutine 都不会获取到读取锁
```

File: src/sync/rwmutex.go [hunk 13]
```diff
@@ -10,36 +10,31 @@ import (
...
-	w           Mutex  // held if there are pending writers
-	writerSem   uint32 // semaphore for writers to wait for completing readers
-	readerSem   uint32 // semaphore for readers to wait for completing writers
-	readerCount int32  // number of pending readers
-	readerWait  int32  // number of departing readers
+	w           Mutex  // writer的互斥锁
+	writerSem   uint32 // writer的信号量
+	readerSem   uint32 // reader的信号量
+	readerCount int32  // reader的数量
+	readerWait  int32  // 进行write需要等待的reader数量
```
Comment (darcyaf) [inline src/sync/rwmutex.go]: 这里的空格最好加一下

File: src/sync/rwmutex.go [hunk 14]
```diff
@@ -10,36 +10,31 @@ import (
...
 
 const rwmutexMaxReaders = 1 << 30
 
-// RLock locks rw for reading.
+// RLock 读取锁锁定.
 //
-// It should not be used for recursive read locking; a blocked Lock
-// call excludes new readers from acquiring the lock. See the
-// documentation on the RWMutex type.
+// 禁止递归调用
```
Comment (darcyaf) [inline src/sync/rwmutex.go]:
```suggestion
// 禁止递归调用. 阻塞中的 Lock 调用会阻止新的 reader 来获取锁.
```

File: src/sync/rwmutex.go [hunk 15]
```diff
@@ -55,10 +50,9 @@ func (rw *RWMutex) RLock() {
 	}
 }
 
-// RUnlock undoes a single RLock call;
-// it does not affect other simultaneous readers.
-// It is a run-time error if rw is not locked for reading
-// on entry to RUnlock.
+// RUnlock 解锁一个RLock锁定的锁;
+// 它不会影响同时并发读的其他goroutine.
+// 如果取解锁一个没有锁定的锁，会导致报错.
```
Comment (darcyaf) [inline src/sync/rwmutex.go]:
```suggestion
// 如果去解锁一个没有锁定的锁，会导致报错.
```

File: src/sync/waitgroup.go [hunk 16]
```diff
@@ -10,21 +10,19 @@ import (
...
-// goroutines to wait for. Then each of the goroutines
-// runs and calls Done when finished. At the same time,
-// Wait can be used to block until all goroutines have finished.
+// WaitGroup 用来等待一组goroutines的结束，.
+// 主goroutine 调用Add方法去设置一共需要等待的数量
+// 每个被等待的goroutine在执行完毕的时候需要执行Done方法来表示完成
+// 同时,调用Wait方法可以阻塞主goroutine直到所有的goroutine完成
 //
-// A WaitGroup must not be copied after first use.
+// WaitGroup 在时候后不允许拷贝.
```
Comment (darcyaf) [inline src/sync/waitgroup.go]:
```suggestion
// WaitGroup 在使用后不允许拷贝.
```

File: src/sync/waitgroup.go [hunk 17]
```diff
@@ -94,12 +88,12 @@ func (wg *WaitGroup) Add(delta int) {
 	}
 }
 
-// Done decrements the WaitGroup counter by one.
+// Done 减少1个等待的counter.
```
Comment (darcyaf) [inline src/sync/waitgroup.go]:
```suggestion
// Done 减少 1 个等待的counter.
```

File: src/sync/waitgroup.go [hunk 18]
```diff
@@ -94,12 +88,12 @@ func (wg *WaitGroup) Add(delta int) {
...
 func (wg *WaitGroup) Done() {
 	wg.Add(-1)
 }
 
-// Wait blocks until the WaitGroup counter is zero.
+// Wait 阻塞goroutine直到需要等待的goroutine数量为0
```
Comment (darcyaf) [inline src/sync/waitgroup.go]:
```suggestion
// Wait 阻塞当前 goroutine 直到等待的 WaitGroup的 counter 的值 为 0
```

**Review by darcyaf — COMMENTED** [review]
有部分文字错误的我提了suggestion, 其他的还有很多英文数字和中文之间缺了空格，这个有空的时候可以改改

**Comment by Code-Fight** [comment]
好了，辛苦再看一下

**Comment by Code-Fight** [comment]
@darcyaf 还有什么问题吗？

**Pull request merged**